pkg/execute/kubectl.go

```diff
@@ -3,51 +3,58 @@ package execute
-	kubectlNotAllowedKindMsgFmt      = "Sorry, the kubectl command is not authorized to work with '%s' resources on cluster '%s'. Use 'commands list' to see all allowed resources."
-	kubectlDefaultNamespace          = "default"
+	kubectlNotAuthorizedMsgFmt         = "Sorry, this channel is not authorized to execute kubectl command on cluster '%s'."
+	kubectlNotAllowedVerbMsgFmt        = "Sorry, the kubectl '%s' command cannot be executed in the '%s' Namespace on cluster '%s'. Use 'commands list' to see allowed commands."
+	kubectlNotAllowedVerbInAllNsMsgFmt = "Sorry, the kubectl '%s' command cannot be executed for all Namespaces on cluster '%s'. Use 'commands list' to see allowed commands."
+	kubectlNotAllowedKindMsgFmt        = "Sorry, the kubectl command is not authorized to work with '%s' resources in the '%s' Namespace on cluster '%s'. Use 'commands list' to see allowed commands."
+	kubectlNotAllowedKinInAllNsMsgFmt  = "Sorry, the kubectl command is not authorized to work with '%s' resources for all Namespaces on cluster '%s'. Use 'commands list' to see allowed commands."
+	kubectlFlagAfterVerbMsg            = "Please specify the resource name after the verb, and all flags after the resource name. Format <verb> <resource> [flags]"
+	kubectlDefaultNamespace            = "default"
 )
 
 // Kubectl executes kubectl commands using local binary.
 type Kubectl struct {
-	log        logrus.FieldLogger
-	cfg        config.Config
-	resMapping ResourceMapping
-	runCmdFn   CommandRunnerFunc
+	log logrus.FieldLogger
+	cfg config.Config
+
+	kcChecker *kubectl.Checker
+	runCmdFn  CommandRunnerFunc
+	merger    *kubectl.Merger
 }
 
 // NewKubectl creates a new instance of Kubectl.
-func NewKubectl(log logrus.FieldLogger, cfg config.Config, mapping ResourceMapping, fn CommandRunnerFunc) *Kubectl {
+func NewKubectl(log logrus.FieldLogger, cfg config.Config, merger *kubectl.Merger, kcChecker *kubectl.Checker, fn CommandRunnerFunc) *Kubectl {
 	return &Kubectl{
-		log:        log,
-		cfg:        cfg,
-		resMapping: mapping,
-		runCmdFn:   fn,
+		log:       log,
+		cfg:       cfg,
+		merger:    merger,
+		kcChecker: kcChecker,
+		runCmdFn:  fn,
 	}
 }
 
 // CanHandle returns true if it's allowed kubectl command that can be handled by this executor.
 //
 // TODO: we should just introduce a command name explicitly. In this case `@BotKube kubectl get po` instead of `@BotKube get po`
 // As a result, we are able to detect kubectl command but say that you're simply not authorized to use it instead of "Command not supported. (..)"
-func (e *Kubectl) CanHandle(args []string) bool {
+func (e *Kubectl) CanHandle(bindings []string, args []string) bool {
 	if len(args) == 0 {
 		return false
 	}
 
 	// Check if such kubectl verb is enabled
-	if !e.resMapping.AllowedKubectlVerbMap[args[0]] {
+	if !e.kcChecker.IsKnownVerb(e.merger.MergeAllEnabledVerbs(bindings), args[0]) {
```

Comment:
Slight nit pick.

What if the merger simply told you if it includes a verb? e.g. `e.merger.includesVerb(args[0]])`

Reply:
IMO it would be good to have that separated. For now, the boundaries are quite easy to describe. Merger takes care of any sort of merging strategies. Checker, on the other hand, it's used to assert some expectation. I'm not sure if it's worth to mix this functionality even if it's simple. I tried to avoid having a one component that will do all stuff, instead wanted to provide smaller components where the logic is simple enough to be well tested and read.

Reply:
I see what you're getting at - I just felt there's a bit of jumping around.

Ensuring the components are decoupled trumps everything. Here, I felt a generic (`includes...`)[tell don't ask](https://martinfowler.com/bliki/TellDontAsk.html) might help.

It definitely should not delay merging this PR - so all good.